### PR TITLE
Add arm64e and pointer authentication support for Swift

### PR DESCRIPTION
**Pull request by kubamracek**
This PR upstreams the internally developed support for arm64e and pointer authentication in Swift. For details about pointer authentication, see <http://lists.llvm.org/pipermail/llvm-dev/2019-October/136091.html>, and for the LLVM and Clang changes, see <https://github.com/apple/llvm-project/pull/14>. This work is primarily authored by John McCall, with many smaller contributions by several others at Apple including myself.

### CODE REVIEW

**Comment by kubamracek** [comment]
Please test with following pull request:
https://github.com/apple/llvm-project/pull/846

@swift-ci Please test

**Comment by kubamracek** [comment]
Please test with following pull request:
apple/llvm-project#846

@swift-ci Please test

**Comment by kubamracek** [comment]
@swift-ci Please test

**Comment by kubamracek** [comment]
Please test with following pull request:
apple/llvm-project#846

@swift-ci Please test

**Comment by rjmccall** [comment]
The IRGen test conflicts should be trivial to fix: in this patch, the 64-bit test line has been split into ptrauth and non-ptrauth variants, so non just copy the new symbol name from the single 64-bit line into both, preserving the `.ptrauth` suffix on the ptrauth line.

**Comment by kubamracek** [comment]
Please test with following pull request:
apple/llvm-project#846

@swift-ci Please test

**Comment by kubamracek** [comment]
@swift-ci Please test

**Comment by kubamracek** [comment]
Please test with following pull request:
apple/llvm-project#846

@swift-ci Please test

**Comment by kubamracek** [comment]
Please test with the following pull request:
https://github.com/apple/llvm-project/pull/863

@swift-ci Please test

**Comment by kubamracek** [comment]
@swift-ci Please test

**Comment by kubamracek** [comment]
Please test with the following pull request:
apple/llvm-project#863

@swift-ci Please test

**Comment by kubamracek** [comment]
@swift-ci Please test

**Comment by kubamracek** [comment]
@swift-ci Please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

File: stdlib/public/runtime/Metadata.cpp [hunk 1]
```diff
@@ -4207,6 +4301,110 @@ static bool doesNotRequireInstantiation(
   return true;
 }
 
+static const unsigned swift_ptrauth_key_associated_type =
```
Comment (smeenai) [inline stdlib/public/runtime/Metadata.cpp]: Was it intentional to assign this variable to itself? Clang accepts this without warning :/

File: stdlib/public/runtime/Metadata.cpp [hunk 2]
```diff
@@ -4207,6 +4301,110 @@ static bool doesNotRequireInstantiation(
...
+                                              const Metadata *witness,
+                                              const ProtocolRequirement &reqt) {
+  assert(reqt.Flags.getKind() ==
+           ProtocolRequirementFlags::Kind::AssociatedTypeAccessFunction);
+  // FIXME: this should use ptrauth_key_process_independent_data
+  // now that it no longer stores a function pointer.
+  swift_ptrauth_init(slot, witness, reqt.Flags.getExtraDiscriminator());
+}
+
+static const unsigned swift_ptrauth_key_associated_conformance =
```
Comment (smeenai) [inline stdlib/public/runtime/Metadata.cpp]: Same here.

**Review by smeenai — COMMENTED** [review]
This broke builds for me:

```
<swift>/stdlib/public/runtime/Metadata.cpp:4645:40: error: expression is not an integer constant expression
  witness = ptrauth_auth_data(witness, swift_ptrauth_key_associated_type,
                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<clang>/usr/lib/clang/11.0.0/include/ptrauth.h:236:35: note: expanded from macro 'ptrauth_auth_data'
  __builtin_ptrauth_auth(__value, __old_key, __old_data)
                                  ^~~~~~~~~
<swift>/stdlib/public/runtime/Metadata.cpp:4767:40: error: expression is not an integer constant expression
  witness = ptrauth_auth_data(witness, swift_ptrauth_key_associated_type,
                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<clang>/usr/lib/clang/11.0.0/include/ptrauth.h:236:35: note: expanded from macro 'ptrauth_auth_data'
  __builtin_ptrauth_auth(__value, __old_key, __old_data)
                                  ^~~~~~~~~
<swift>/stdlib/public/runtime/Metadata.cpp:4818:26: error: expression is not an integer constant expression
                witness, swift_ptrauth_key_associated_conformance,
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<clang>/usr/lib/clang/11.0.0/include/ptrauth.h:236:35: note: expanded from macro 'ptrauth_auth_data'
  __builtin_ptrauth_auth(__value, __old_key, __old_data)
                                  ^~~~~~~~~
<swift>/stdlib/public/runtime/Metadata.cpp:4894:40: error: expression is not an integer constant expression
  witness = ptrauth_auth_data(witness, swift_ptrauth_key_associated_conformance,
                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<clang>/usr/lib/clang/11.0.0/include/ptrauth.h:236:35: note: expanded from macro 'ptrauth_auth_data'
  __builtin_ptrauth_auth(__value, __old_key, __old_data)
                                  ^~~~~~~~~
4 errors generated.
```

File: stdlib/public/runtime/Metadata.cpp [hunk 3]
```diff
@@ -4207,6 +4301,110 @@ static bool doesNotRequireInstantiation(
   return true;
 }
 
+static const unsigned swift_ptrauth_key_associated_type =
+  swift_ptrauth_key_associated_type;
```
Comment (smeenai) [inline stdlib/public/runtime/Metadata.cpp]: I don't see any way to expand the context around my previous comment, so this is where the self-assignment is happening.

**Comment by smeenai** [comment]
Ah, I see https://github.com/apple/swift/pull/30363.

**Pull request merged**